### ISSUE TITLE
LaTeX builder use of amsfonts may break user choice of math fonts

### ISSUE THREAD

**Issue by jfbu** [sphinx-doc/sphinx#2290]
Commit 3543caf2140f25ceefccc7c6fb4b9fa6dce180db which Merge pull request #2186 from @xuhdev for support of `\mathbb` may break user choices made via the `latex_elements` dictionary.

Indeed `\usepackage{amsfonts}` end up after the user customizable `'fontpkg'` key to `latex_elements`. But this is where the user may wish to specify some math font packages; this will be at best overwritten by `\usepackage{amsfonts}`, but may raise errors, for example I got

``` latex
LaTeX Error: Command `\ulcorner' already defined.
```

because I was issuing `\setmathfont[Extension = .otf,BoldFont = *bold]{xits-math}` as part of `'fontpkg'` key (this is with lualatex compilation, and `\usepackage{unicode-math}` in an earlier key).

Fix for me was to issue `\usepackage{amsfonts}` before (I used  `'fontenc'` key) so it is already loaded, then overwritten by `XITS Math`, then the second `\usepackage{amsfonts}` is harmless because LaTeX will not load it a second time.

I propose that `amsfonts` be loaded earlier, or not all, depending on a user boolean perhaps. Don't know Sphinx internals enough to figure out the best fix. Perhaps this could be made engine dependent: in case of lualatex or xelatex, don't load automatically `amsfonts` which is for traditional, not OpenType math fonts.

**Comment by xuhdev** [comment]
@tk0miya We can either revert my commit and advice users to add the package in `latex_elements` in the doc for math extension, or add a boolean conf option for the math extension to not load this package. Which one do you think is better? I prefer the second solution as it is so common.

**Comment by tk0miya** [comment]
@xuhdev What about loading `amsfonts` at earlier step?
If good enough for pdflatex, lualatex and xelatex, it looks good solution for me.

Or not, it should not load it by default. because the dependencies between latex package is very difficult for me...

**Comment by xuhdev** [comment]
@tk0miya It should work if loading it at an earlier step, but I think the original poster of this issue also points out that luatex does not need `amsfonts`---it can use OpenType math fonts.

**Comment by jfbu** [comment]
Users of lualatex/xelatex will wish to use package `unicode-math`.

Its documentation says it is better if `amsmath` is loaded before it; regarding `amsfonts` however I tested and it seems that `unicode-math` undoes `amsfonts` dealings, even if `amsfonts` is loaded after it.

``` latex
\documentclass{article}

\usepackage{unicode-math}
\usepackage{amsfonts}
\begin{document}
\showoutput % check the log for the line
            % ....\EU2/latinmodern-math.otf(0)/m/n/10 ℤ
            % (with lualatex; with xelatex it is EU1 and glyph#3518 rather
            % than ℤ)
            % (with future TeXLive 2016, EU1/EU2 will be replaced by TU)
% This shows that classic 256 slots amsfont for blackboard bold has not been
% used, despite amsfonts being loaded after unicode-math.
% 
% Besides, \mathbb works without amsfonts being loaded.
$$\mathbb{Z}$$
\end{document}
```

LaTeX log contains:

```
....\EU2/latinmodern-math.otf(0)/m/n/10 ℤ
```

if document is compiled with lualatex, independently of whether `amsfonts` is loaded before, after, or not at all. With `unicode-math` there is no need for `amsfonts` for `\mathbb` or `\mathfrak`.

Note that `unicode-math` is only for lualatex or xelatex compilation, it will not work with pdflatex.

**Comment by xuhdev** [comment]
OK, then I agree to put amsfonts to an earlier part.

**Comment by tk0miya** [comment]
@jfbu @xuhdev 
I fixed the order of usepackages at 532e99c (stable).
If you find any problems, please let me know.

Thanks,

**Comment by jfbu** [comment]
looks good to me.
